client/src/client.rs

```diff
@@ -197,20 +198,16 @@ pub enum Auth {
 impl Auth {
     /// Convert into the arguments that jsonrpc::Client needs.
     pub fn get_user_pass(self) -> Result<(Option<String>, Option<String>)> {
-        use std::io::Read;
         match self {
             Auth::None => Ok((None, None)),
             Auth::UserPass(u, p) => Ok((Some(u), Some(p))),
-            Auth::CookieFile(path) => {
-                let mut file = File::open(path)?;
-                let mut contents = String::new();
-                file.read_to_string(&mut contents)?;
-                let mut split = contents.splitn(2, ":");
-                Ok((
-                    Some(split.next().ok_or(Error::InvalidCookieFile)?.into()),
-                    Some(split.next().ok_or(Error::InvalidCookieFile)?.into()),
-                ))
-            }
+            Auth::CookieFile(path) => BufReader::new(File::open(path)?)
+                .lines()
+                .next()
+                .ok_or(Error::InvalidCookieFile)??
+                .split_once(':')
+                .map(|(user, pass)| (Some(user.into()), Some(pass.into())))
+                .ok_or(Error::InvalidCookieFile),
```

Comment:
So if you have a cookie file like this:

```
foo:bar
some
garbage
here
```

This will ignore the additional garbage. Is it actually the behavior of Core? Is it even desirable to silently ignore it? It could lead to significant debugging nightmares.

I'd personally just call `trim_end_matches('\n')` just before `split`. It only removes the newline at the end and is a simpler change. Or maybe check that there's only one line and error if there are more.

Reply:
Agreed it's weird, but it is the behavior of core. This is the function in Core:

https://github.com/bitcoin/bitcoin/blob/e09ad284c762a79d59417389e9056c18e25d9770/src/rpc/request.cpp#L113-L127

It calls std::getline on the cookiefile, which returns one line, and ignores anything else.

Overall, I think accepting the same cookie files that core accepts is probably good. I actually ran into this, where I had a cookie file that was accepted by bitcoin-cli, but not by rust-bitcoin.

Reply:
Yeah, being consistent is good. I think both Core and this library should at least emit a warning if there's more than one line though.

Reply:
That might be nice, it would be a slightly larger change though.

Reply:
Would need an optional `log` dependency for sure. Checking is actually easy, just test `lines.next().is_some()`